src/ray/core_worker/reference_count.cc

```diff
@@ -939,6 +939,38 @@ void ReferenceCounter::HandleObjectSpilled(const ObjectID &object_id) {
   ReleasePlasmaObject(it);
 }
 
+absl::optional<LocalityData> ReferenceCounter::GetLocalityData(
+    const ObjectID &object_id) {
+  absl::MutexLock lock(&mutex_);
+  // Uses the reference table to return locality data for an object.
+  auto it = object_id_refs_.find(object_id);
+  if (it != object_id_refs_.end()) {
+    // Reference for object_id exists.
+    if (auto node_id = it->second.pinned_at_raylet_id) {
+      // The raylet ID to which this reference is pinned is defined.
+      auto object_size = it->second.object_size;
+      if (object_size >= 0) {
+        // The object size of this reference is known.
+        absl::optional<LocalityData> locality_data(
+            {static_cast<uint64_t>(object_size), {node_id.value()}});
+        return locality_data;
+      } else {
+        RAY_LOG(DEBUG) << "Reference " << it->second.call_site << " for object "
+                       << object_id
+                       << " has an unknown object size, locality data not available";
+      }
+    } else {
+      RAY_LOG(DEBUG)
+          << "Reference " << it->second.call_site << " for object " << object_id
+          << " doesn't have a defined pinned raylet ID, locality data not available";
+    }
+  } else {
+    RAY_LOG(DEBUG) << "Object " << object_id
+                   << " not in reference table, locality data not available";
+  }
+  return absl::nullopt;
```

Comment:
I think we just return {} for empty optional value in other parts of repo.

Reply:
I did see that done elsewhere in the repo, and I'm happy to adhere to the current practice of relying on default initialization of `absl::optional<T>` in these kind of "return an unset optional" cases, but I do think there are a few advantages of returning `absl::nullopt` instead of `{}`:

1. I personally think that it's more readable, and that it more clearly conveys that we're returning an unset optional. It may be because my C++ is a bit rusty, but I had to think through the `{}` return to make sure that this default initialization did what I expected, while with the `absl::nullopt` return, the semantics were immediately clear to me.
2. There's been [some weirdness in the past](https://stackoverflow.com/questions/57964217/stdoptional-construct-empty-with-or-stdnullopt) where default initialization ends up value-initializing (i.e. zero-initializing) `optional` implementations that don't have a user-provided default constructor (i.e. the default constructor is implicitly declared or explicitly defaulted), causing unexpected overhead for something that people _thought_ was equivalent to the `absl::nullopt` constructor path. It was made worse by this happening in GCC but not Clang. This resulted in the recommendation of returning an explicit `absl::nullopt` instead, to avoid that ambiguity/weirdness.

(2) has been [fixed in GCC](https://github.com/gcc-mirror/gcc/commit/bafd12cb22e83b7da8946873513a897e48e2900f#diff-98ccb1166278152524d4e41464c3c17597ca7409fcee5756759f900359838dc5), Clang's implementation of `std::optional` has (I think) always had [a user-provided default constructor](https://github.com/llvm/llvm-project/blob/9c978dd6e12e5ffaf5441f459db47e9892b09a82/libcxx/include/optional#L688), and I've checked the implementation of `absl::optional` and it [seems as if this isn't an issue](https://github.com/abseil/abseil-cpp/blob/c5c4db4f5191fe5e76cbf68dcc71fb28702f7d2b/absl/types/optional.h#L128-L130), but I still decided to go with `absl::nullopt` since it seemed more explicit and readable.

(2) aside, let me know if you think returning `absl::nullopt` is more readable than returning `{}`, and maybe we can port to that pattern in the future. In the meantime, I'll stick with returning `{}` like other uses of `absl::optional` in the repo. 😄

Reply:
I actually agree nullopt is more readable. It is probably a good idea to just use nullopt instead of {} everywhere in this case.

Reply:
Ah, cool! I'll change them back, and we can try to evangelize the `absl::nullopt` pattern elsewhere in the repo in the future.